### ISSUE TITLE
Remove dependencies to the UI from the non-UI bundles.

### ISSUE THREAD

**Issue by yoosiba** [eclipse/n4js#100]
Dependency tree in N4JS is currently convoluted in regards of UI and non-UI bundles.
Best example of that is `org.eclipse.n4js.utils` -> `org.eclipse.xtext.ui` -> `org.eclipse.ui` -> ...
The `org.eclipse.n4js.utils` bundle is very high in our dependency graph as `org.eclipse.n4js` (the main bundle) depends on it as well as other bundles (e.g. `org.eclipse.n4js.ts`) that have nothing to do with the UI.

### N4JSC.jar implications
Headless compiler is composed of bundles that (due to problems described above) depend on UI. Since build of that tool is handcrafted, we build it without including those transitive dependencies. This means that at any given time N4JSC.jar can crash with `ClassNotFoundException` or `NoClassDefFoundError` if only some code path leads to UI bundle.

This also blocks #96 as OSGI environemnt will validate bundle dependencies at start / install and will throw exceptions if validation will be missing. See example resolution below 

Example dependency resolution by the OSGI environment
```
!ENTRY org.eclipse.n4js.hlc.base 4 0 2017-07-21 16:41:04.590
!MESSAGE FrameworkEvent ERROR
!STACK 0
org.osgi.framework.BundleException: Could not resolve module: org.eclipse.n4js.hlc.base [247]
  Unresolved requirement: Require-Bundle: org.eclipse.n4js.ts
    -> Bundle-SymbolicName: org.eclipse.n4js.ts; bundle-version="0.0.1.qualifier"; singleton:="true"
       org.eclipse.n4js.ts [316]
         Unresolved requirement: Require-Bundle: org.eclipse.n4js.utils
           -> Bundle-SymbolicName: org.eclipse.n4js.utils; bundle-version="0.0.1.qualifier"; singleton:="true"
              org.eclipse.n4js.utils [421]
                Unresolved requirement: Require-Bundle: org.eclipse.xtext.ui
                  -> Bundle-SymbolicName: org.eclipse.xtext.ui; bundle-version="2.12.0.v20170519-0809"; singleton:="true"
                     org.eclipse.xtext.ui [332]
                       Unresolved requirement: Require-Bundle: org.eclipse.ui; bundle-version="3.4.0"; visibility:="reexport"
                         -> Bundle-SymbolicName: org.eclipse.ui; bundle-version="3.108.1.v20160929-1045"; singleton:="true"
                            org.eclipse.ui [410]
                              Unresolved requirement: Require-Bundle: org.eclipse.ui.workbench; bundle-version="[3.105.0,4.0.0)"; visibility:="reexport"
                                -> Bundle-SymbolicName: org.eclipse.ui.workbench; bundle-version="3.108.2.v20161025-2029"; singleton:="true"
                                   org.eclipse.ui.workbench [379]
                                     Unresolved requirement: Import-Package: org.eclipse.e4.ui.internal.workbench.renderers.swt
                                       -> Export-Package: org.eclipse.e4.ui.internal.workbench.renderers.swt; bundle-symbolic-name="org.eclipse.e4.ui.workbench.renderers.swt"; bundle-version="0.14.0.v20160525-0940"; version="0.0.0"; x-friends:="org.eclipse.ui.workbench"
                                          org.eclipse.e4.ui.workbench.renderers.swt [446]
                                            Unresolved requirement: Require-Bundle: org.eclipse.e4.ui.workbench.swt; bundle-version="0.9.0"
                                              -> Bundle-SymbolicName: org.eclipse.e4.ui.workbench.swt; bundle-version="0.14.1.v20160829-0832"; singleton:="true"
                                                 org.eclipse.e4.ui.workbench.swt [376]
                                                   Unresolved requirement: Require-Capability: osgi.extender; filter:="(&(osgi.extender=osgi.component)(version>=1.2)(!(version>=2.0)))"
                                     Unresolved requirement: Import-Package: org.eclipse.e4.ui.internal.workbench.swt
                                       -> Export-Package: org.eclipse.e4.ui.internal.workbench.swt; bundle-symbolic-name="org.eclipse.e4.ui.workbench.swt"; bundle-version="0.14.1.v20160829-0832"; version="0.0.0"; x-friends:="org.eclipse.e4.ui.workbench.addons.swt,org.eclipse.e4.ui.workbench.renderers.swt,org.eclipse.ui.workbench"
(...)
  Unresolved requirement: Require-Bundle: org.eclipse.n4js.n4mf
    -> Bundle-SymbolicName: org.eclipse.n4js.n4mf; bundle-version="0.0.1.qualifier"; singleton:="true"
       org.eclipse.n4js.n4mf [380]
         Unresolved requirement: Require-Bundle: org.eclipse.n4js.utils
           -> Bundle-SymbolicName: org.eclipse.n4js.utils; bundle-version="0.0.1.qualifier"; singleton:="true"
  Unresolved requirement: Require-Bundle: org.eclipse.n4js
    -> Bundle-SymbolicName: org.eclipse.n4js; bundle-version="0.0.1.qualifier"; singleton:="true"

	at org.eclipse.osgi.container.Module.start(Module.java:444)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1620)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1599)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1571)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1514)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
```

### Hint
One way of testing dependency graph from N4JSC point of view, without solving #96 completely is to create product definition for hlc and then telling eclipse to validate the product.

**Comment by yoosiba** [comment]
Another example
```
org.eclipse.n4js
  -> org.eclipse.xtend.builder
        -> org.eclipse.ui
```

**Comment by yoosiba** [comment]
Another example 
```
org.eclipse.n4js.generator.headless
   -> org.eclipse.n4js.environments
           -> org.eclipse.xtext.ui
                    -> org.eclipse.ui
           -> org.eclipse.builder
                    -> org.eclipse.ui
```

**Comment by yoosiba** [comment]
merged  f182db9
